Issue #7183: add JavadocMissingWhitespaceAfterAsteriskCheck

0blivious opened the pull request:
Issue #7183

Contribution PR:
checkstyle/contribution#454
https://github.com/checkstyle/contribution/pull/454 (this link was added to satisfy `verify-no-exception-configs` build item)

Regression report:
https://0blivious.github.io/7183/index.html

Old regression report
https://0blivious.github.io/index.html

Sevntu-checkstyle PR (merged)
https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/806

Checkstyle PR
https://github.com/checkstyle/checkstyle/pull/7883

rnveach commented:
> https://0blivious.github.io/index.html

Please look over report and configure it for the check you are making.
https://0blivious.github.io/apache-jsecurity/index.html
You are not testing regression on `ThrowsCount`.
https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#configuration-file

0blivious commented:
> Please look over report and configure it for the check you are making.
> https://0blivious.github.io/apache-jsecurity/index.html
> You are not testing regression on `ThrowsCount`.
> https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#configuration-file

I have updated the report. The CI is failing as there are other files in CheckStyle that violate this check. Can tell me how to work around this? Thx very much.

rnveach commented:
>  The CI is failing as there are other files in CheckStyle that violate this check. Can tell me how to work around this?

If the violations are correct, you must fix the issues in Checkstyle's code base.

romani commented:
Please squash all commits in one and resolve All CIs, all CIs should be green. 
If you have problems let us know

0blivious commented:
> Please squash all commits in one and resolve All CIs, all CIs should be green.
> If you have problems let us know

Hi before I correct all the missing whitespace in CheckStyle code I would like to check, should this check apply to single line comment?

```
/**This ok or violation? */
```

rnveach commented:
@0blivious Why wouldn't the check apply to all javadocs and all asterisks in the javadoc?

pbludov commented:
> Hi before I correct all the missing whitespace in CheckStyle code I would like to check, should this check apply to single line comment?

Yes, this should be violation.

pbludov commented:
From https://0blivious.github.io/apache-struts/xref/nus/sem6/contribution/checkstyle-tester/repositories/apache-struts/core/src/main/java/com/opensymphony/xwork2/util/ResolverUtil.java.html#L58
```java
58   *<pre>
59   *ResolverUtil&lt;ActionBean&gt; resolver = new ResolverUtil&lt;ActionBean&gt;();
60   *resolver.findImplementation(ActionBean.class, pkg1, pkg2);
61   *resolver.find(new CustomTest(), pkg1);
62   *resolver.find(new CustomTest(), pkg2);
63   *Collection&lt;ActionBean&gt; beans = resolver.getClasses();
64   *</pre> 
```
The violation occurs only on the first line.

@0blivious, please find out what is the output of the Javadoc tool for this class? And DetailAst generated with `checkstyle -J`
See reference [here](https://checkstyle.org/cmdline.html#Command_line_usage).

0blivious commented:
> ```java
> 58   *<pre>
> 59   *ResolverUtil&lt;ActionBean&gt; resolver = new ResolverUtil&lt;ActionBean&gt;();
> 60   *resolver.findImplementation(ActionBean.class, pkg1, pkg2);
> 61   *resolver.find(new CustomTest(), pkg1);
> 62   *resolver.find(new CustomTest(), pkg2);
> 63   *Collection&lt;ActionBean&gt; beans = resolver.getClasses();
> 64   *</pre> 
> ```
> 
> The violation occurs only on the first line.

I have changed the check implementation and updated the regression report. It now will report violation for all these lines. I got some new problem regarding the implementation with comment like those in https://0blivious.github.io/Hbase/index.html
```
    /*********
     * State
     *********/
```
The DetailAst will be
```
JAVADOC -> JAVADOC
|--LEADING_ASTERISK -> *
|--TEXT ->  ******
|--NEWLINE -> \n
|--LEADING_ASTERISK -> *
|--TEXT ->  State
|--NEWLINE -> \n
|--LEADING_ASTERISK -> *
|--TEXT ->  *******
|--EOF -> <EOF>
```
The current `visitJavadocToken` will report violation for the 1st and 3rd line. Should we tolerate this type of comment?

Also can I get some help with the pitest2? I am not sure why it fails......Thx in advance!

pbludov commented:
>     /*********
>      * State
>      *********/

There should be no violation. The html generated by the Javadoc tool is `State` (no asterisks at all).
Leading asterisks are tricky and processed in non obvious way.
The [specs](https://docs.oracle.com/en/java/javase/11/docs/specs/doc-comment-spec.html) says that there may be many leading asterisks on the line. Our Javadoc parser is more strict: all asterisks except the first one are treated as a text. So, you have to process the text node after the leading asterisk and skip all `*` on the front.  For the case
```
|--LEADING_ASTERISK -> *
|--TEXT ->  ******
|--NEWLINE -> \n
```
it should be same as
```
|--LEADING_ASTERISK -> *
|--NEWLINE -> \n
```

May be this check should be a Java check, not Javadoc check, as [this](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocContentLocationCheck.java) one.

pbludov commented:
@0blivious ping. Please let us know if you have a problem with the issue.

0blivious commented:
> @0blivious ping. Please let us know if you have a problem with the issue.

Sorry was a bit busy for these few days. I have updated the implementation to skip * in front. I have some problems with the CI though. Can tell me how to interpret the pitest result and fix it?

rnveach commented:
> Can tell me how to interpret the pitest result and fix it?

Please see https://github.com/checkstyle/checkstyle/issues/7797 on an introduction to pitest and some helpful links.

pbludov commented:
@0blivious Pitest fails with
```bash
Diff:
--- target/ignored.txt	2020-03-12 18:29:01.267472496 +0000
+++ target/actual.txt	2020-03-12 18:29:01.263472490 +0000
@@ -1,5 +1,4 @@
 AbstractJavadocCheck.java.html:<td class='covered'><pre><span  class='survived'>            Arrays.sort(acceptableJavadocTokens);</span></pre></td></tr>
-AbstractJavadocCheck.java.html:<td class='covered'><pre><span  class='survived'>            Arrays.sort(defaultJavadocTokens);</span></pre></td></tr>
 AbstractJavadocCheck.java.html:<td class='covered'><pre><span  class='survived'>        beginJavadocTree(root);</span></pre></td></tr>
 AbstractJavadocCheck.java.html:<td class='covered'><pre><span  class='survived'>        finishJavadocTree(root);</span></pre></td></tr>
 AbstractJavadocCheck.java.html:<td class='covered'><pre><span  class='survived'>        javadocTokens.clear();</span></pre></td></tr>
Difference between 'Actual' and 'Ignore' lists is detected, lists should be equal.
build will be failed.
```

This means that this PR makes the suppression https://github.com/checkstyle/checkstyle/blob/master/.ci/pitest.sh#L184 redundant. Please remove this line from `pitest.sh`.

This check has multiple Javadoc tokens and they are sorted in alphabetical order, but this order does not match the order of tag values:
`LEADING_ASTERISK=1`, `JAVADOC=10000`

pbludov commented:
@0blivious please send a PR to fix sevntu-checkstyle [issues](https://0blivious.github.io/sevntu-checkstyle/index.html) to https://github.com/sevntu-checkstyle/sevntu.checkstyle/

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheck.java:
```diff
@@ -0,0 +1,123 @@
+
+/**
+ * <p>
+ * Checks that at there is least one whitespace after the leading asterisk.
```
that at there is least -> that there is at least

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheck.java:
```diff
@@ -0,0 +1,123 @@
+ * <pre>
+ * class TestClass {
+ *   &#47;**
+ *     *This is invalid java doc.
```
java doc -> Javadoc
(here and below)

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheck.java:
```diff
@@ -0,0 +1,123 @@
+ *     *&#47;
+ *   void validJavaDocMethod() {
+ *   }
+ *   &#47;**This is invalid single linejava doc. *&#47;
```
invalid single linejava doc -> invalid single-line Javadoc

pbludov commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,53 @@
+    /***/
+    void bar4() {}
+
+    /** This is ok. **/
```
Please add a test case for
`/** *violation */`
A single-line Javadoc with leading asterisk. Yes, it is weird, but it is legal.

0blivious replied:
done

pbludov commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,59 @@
+    void foo() {}
+
+    /**
+     *  @see See somthing.
```
somthing -> something

pbludov commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,59 @@
+    void foo1() {}
+
+    /**
+     *This is not ok.
```
Please change all `not ok ` with `violation`

pbludov commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,59 @@
+    void foo2() {}
+
+    /**
+     * This is not ok.
```
This is confusing. Should it be
```
/**
 * This is ok.
 *@see violtaion
 */
```
? Or maybe `There is a violation on the next line` ?

0blivious replied:
done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheckTest.java:
```diff
@@ -0,0 +1,80 @@
+    protected String getPackageLocation() {
+        return "com/puppycrawl/tools/checkstyle/checks/javadoc"
+                +
+                "/javadocmissingwhitespaceafterasterisk";
```
Please combine these 2 lines.

rnveach commented on src/xdocs/config_javadoc.xml:
```diff
@@ -847,6 +847,116 @@ public void doSomething9(File file) throws IOException {
+      <p>Since Checkstyle 8.31</p>
+      <subsection name="Description" id="JavadocMissingWhitespaceAfterAsterisk_Description">
+        <p>
+          Checks that there is at least one whitespace after the leading asterisk.
```
Please add some type of reasoning for why this check's violation matters.

rnveach commented on src/xdocs/config_javadoc.xml:
```diff
@@ -847,6 +847,116 @@ public void doSomething9(File file) throws IOException {
+  void validJavaDocMethod() {
+  }
+  /**This is invalid single-line Javadoc. */
+  void InvalidSingleLineJavaDocMethod() {
```
Please create a valid version.

rnveach commented:
@romani Please finalize.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,59 @@
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocmissingwhitespaceafterasterisk;
+class InputJavadocMissingWhitespaceAfterAsterisk
```
please add 3 lines of empty lines before class we will reuse this in future when Check become to have properties and we need to put in that lines config.

Please do same for other inputs.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsterisk.java:
```diff
@@ -0,0 +1,59 @@
+    void foo1() {}
+
+    /**
+     *Violation.
```
please create Input that has javadocs in most possible places (class, methods, fileds, enums, inner classes, ) and make all cases valid. It will be clear definiton of what is expected as correct.
Please create another Input file (class, methods, fields, ..... ) that will contains invalid cases.

rnveach commented:
@0blivious separate PR was merged. Please rebase and make CI green.

romani commented:
@0blivious , please reply "done" to each my point to make sure nothing is missed.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsteriskInvalid.java:
```diff
@@ -0,0 +1,62 @@
+
+
+/**
+ *Some Javadoc.
```
`Some Javadoc` ==> `violation`.

please do the same for all, to clearly define line that will have violation.
Please apply this to all other cases.

0blivious replied:
done

0blivious commented:
> please regenerate report , there should be no violations - https://0blivious.github.io/sevntu-checkstyle/index.html
> 
> https://0blivious.github.io/Orekit/index.html#A36 valid violation
> `#A36 | warning | JavadocMissingWhitespaceAfterAsterisk | Missing a whitespace after the leading asterisk. | 52 | 8`
> , but .... take a look at source file
> why there is no violation at line 55:

I have regenerated the report and it report violation at line 55 correctly. However, the sevntu-checkstyle still report the same violation. It seems that it still test the old code. I am not sure why this happen......

rnveach commented:
@0blivious If you are referring to the diff script regression, it does not update the repositories automatically. You will have to do this manually.
https://github.com/checkstyle/contribution/issues/180

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmissingwhitespaceafterasterisk/InputJavadocMissingWhitespaceAfterAsteriskInvalid.java:
```diff
@@ -0,0 +1,62 @@
+    /**
+     *<pre>
+     *   Some Javadoc with violation.
+     *</pre>
```
please place word violation on each line where violation is present.
it should be clear where exactly violation be placed , without reading of UT expectations.

in other words: each line at `final String[] expected = {` should have word `violation` in input file, on same line.

0blivious replied:
done and report updated

romani commented:
@0blivious , please remove all git clones from chekcstyle-tester to let him get latest  sources.

0blivious commented:
@romani Hi I have updated the report and fix the input file. Can I get your review on this?

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheck.java:
```diff
@@ -0,0 +1,134 @@
+     * @return true if there is at least one white space after asterisk.
+     */
+    private static boolean hasWhitespaceAfterAsterisk(DetailNode node) {
+        boolean hasWhitespaceAfterAsterisk = true;
```
please change a logic a bit.
At the beginning we do not know that we have whitespace so `hasWhitespaceAfterAsterisk = false;` if you find it you do `hasWhitespaceAfterAsterisk = true;`

0blivious replied:
I set the initial value to be true because we wish to allow comment like
```
 /***   textNode = '*'
  * Something
  */
```
So if the node text is a pure sequence of '\*' I want to return true. If the initial value is set to be false, I will need additional code to perform this check. Is this ok? Or I actually should write another method to check this?

romani replied:
You should never "lie" in code. If you not sure - you are not sure.
At the beginning is in FALSE, if met space it become - true. If no characters other then "*" spacial state that you should consider same as there is space - it has to written in code. Nobody need to remember this.
name method `hasWhitespaceAfterAsteriskBeforeText` and write logic to follow a name.

0blivious replied:
done

romani commented:
@0blivious , please rebase on latest master to resolve wercker issue.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMissingWhitespaceAfterAsteriskCheck.java:
```diff
@@ -0,0 +1,141 @@
+        if (textNode != null
+                && textNode.getType() != JavadocTokenTypes.EOF
+                && !hasWhitespaceAfterAsteriskBeforeText(textNode)) {
+            log(textNode.getLineNumber(), textNode.getColumnNumber(), MSG_KEY);
```
unfortunately I still do not like design .....
name `hasWhitespaceAfterAsteriskBeforeText` is not same as logic inside, it will be a problems in future.

please do:
```
if (textNode != null
  && textNode.getType() != JavadocTokenTypes.EOF) {
    int lastAsterisPosition = getLastLeadingAsterik(....)
    if (!isLast(lastAsterisPosition) && !isWhitespace(lastAsterisPosition +1  )) {
      log(textNode.getLineNumber(), textNode.getColumnNumber(), MSG_KEY);
    }
}
```

and please add tests case with tabs after leading asterik.
Please create new Input file for this as tabs are forbidden in java files in our repo, so we need to suppress violation on it.

0blivious replied:
done

romani replied:
> Please create new Input file for this as tabs are forbidden in java files in our repo, so we need to suppress violation on it.

sad leftover #8019, we do not validate such code for bad whitespaces.
